### PR TITLE
Serialization of custom fields

### PR DESCRIPTION
**Pull request by choinskib**

### CODE REVIEW

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:113]: I think the use of `+` to concatenate strings isn't very idiomatic, and the performance implications aren't as clear as some of the alternatives. Would be good to do something different here.

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:191]: You're not getting line wrapping from rustfmt here because you're inside a macro, but would be good to make sure lines are still wrapped after 100 columns.

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:166]: It's not obvious to me why we need to track missing prefixes here?

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:165]: This way of creating a `proc_macro2::TokenStream` seems much more complex than needed. Also I think the code will look cleaner if you import `proc_macro2::TokenStream` as `TokenStream` and make explicitly qualify `proc_macro::TokenStream`.

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:188]: Shouldn't need the type annotation here.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:42]: We should find a way to do this that doesn't involve allocating a separate string for `self.to_string()`.

File: instant-xml/tests/all.rs
Comment (djc) [inline instant-xml/tests/all.rs:64]: I think we want to ditch the `<test>` tag here and only keep `<Nested>`.

File: instant-xml/tests/all.rs
Comment (djc) [inline instant-xml/tests/all.rs:78]: No need for the `assert_eq!()` here if the `unwrap()` is going to panic.

**Review by djc — COMMENTED** [review]
Here's a round of review on this first PR. I think I'll hold off on reviewing the other ones until we get this one merged.

**Comment by choinskib** [comment]
> BTW, this seems to target the `extended-namespaces` branch but I don't see a PR for that branch?

It is already merged #4, I will rebase this branch to main later on.

**Comment by djc** [comment]
> > BTW, this seems to target the `extended-namespaces` branch but I don't see a PR for that branch?
> 
> It is already merged #4, I will rebase this branch to main later on.

Okay, please rebase it before my next round of review. 👍

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs]: done

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs]: done

File: instant-xml/tests/all.rs
Comment (choinskib) [reply instant-xml/tests/all.rs:64]: done

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs:257]: It is not valid anymore, as I changed the "+" with `push_str`, so it is formatted earlier during macro processing, not inside.

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs]: Changed to `push_str`

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs:230]: Missing_prefix means that there is no matching prefix at current level and it needs to be looked in higher levels (parent, grand parent, etc).  Later in `fn to_xml` we are chacking if prefix exist in higher levels and currently panic if not. Missing prefixes are the ones difined on the fields, we don't need the "not missed" ones, because we can confirm on compile time that those exist. We only need "missed" to check in runtime.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:19]: I don't think this needs the `parent_prefixes` argument, that's only necessary for the "internal" `write_xml()` method.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:16]: Let's wrap `prefixes` in an opaque `Serializer` type and make it required. `Serializer` should implement `Default` so that callers can just pass `Serializer::default()` (instead of `None`).

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:43]: We don't need to provide `to_xml()` here, since the trait provides a default implementation.

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:118]: We should not be comparing to the type name here. Instead we should allow the type's `ToXml` implementation to decide whether it wants to use the field name. So I guess we need to pass the field name into `write_xml()` as part of the serializer state.

File: instant-xml-macros/src/lib.rs
Comment (djc) [reply instant-xml-macros/src/lib.rs]: In instant-xml-macros we still have all this code inside the `quote!()` macro, so it doesn't get formatted as usual IIRC.

Comment (choinskib) [reply instant-xml-macros/src/lib.rs]: Hmm, I did some tests and when I'm expanding the code with `cargo expand` it is formatted correctly. I will dive deeper into this.

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs]: It is moved to `Serializer` struct

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs]: I changed it a little bit, please take a look

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs]: done, please take a look, I rebuilt the whole thing, seems simpler now

**Review by djc — COMMENTED** [review]
This is looking much better, some questions left!

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs]: Right! Thanks for that 😄.

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs]: I removed it, looks better now

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs:27]: The idea is that you just modify the field_data if you want to customize the tag name, without the creating the new value.
In case of the custom type, by default it is taking the struct type as the xml tag (here)[https://github.com/InstantDomain/instant-xml/blob/349af40add28a0bf12142055355f3d1c754f32ec/instant-xml-macros/src/lib.rs#L219]
But if you would like to overide the serialize fn for some struct, you could just modify the `name` inside `FieldData` .

Comment (djc) [reply instant-xml/src/lib.rs:27]: Who would (or, where would you) modify the `field_data`? I'm not sure this idea makes sense. The `field_data` should be exclusively used by a parent type to pass some context into a child type.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:39]: I think we should return an error in the case a scalar type gets called with `None` `FieldData`.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:86]: Let's inline the `match` expression from above here.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:168]: Let's rename `FieldData` to `FieldContext` and rename the `field_` prefixes from the field names (since these are kinda redundant anyway).

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:48]: Should use `write!()` or something to write into the output instead of allocating a `String`, then writing it into the output.

File: instant-xml/src/lib.rs
Comment (djc) [inline instant-xml/src/lib.rs:14]: I think we should keep the `W: fmt::Write` argument and put it in the `Serializer`. How painful is it if the `Serializer` has a generic type argument `W` for this? My experience from Askama is that virtual calls at this level can lead to pretty painful performance.

File: instant-xml-macros/src/lib.rs
Comment (djc) [inline instant-xml-macros/src/lib.rs:231]: We statically now `current_prefixes` so I don't think we have to create a list with them in order to be able to remove them? The loop below could just replace `to_remove` with an array expansion of the current prefixes.

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs:27]: Like this:
`
#[derive(ToXml)]
struct Foo {
    bar: Bar,
}

struct Bar {
    flag: bool,
}

impl ToXml for Bar {
       fn serialize(.., field_data: ..) -> (...) {
                 field_data.unwrap().name = "Something"; // field_name passed from parent is "bar"
                 ...
       }
}
`
But I don't have a strong opinion about that, we can change it to immutable.

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs:39]: It was meant to be used for tuple structs, but I don't know if we are planning to support them.

File: instant-xml-macros/src/lib.rs
Comment (choinskib) [reply instant-xml-macros/src/lib.rs:231]: I might not follow, but the `current_prefixe` may hide the `parent_prefix`. This is the reason for remove, you don't want to remove prefix that existed in parent, we only have to remove the added ones. If there is simpler way of doing that then great, but nothing came to my mind.

File: instant-xml/src/lib.rs
Comment (choinskib) [reply instant-xml/src/lib.rs:14]: I changed it, but I do not get the performance implications here, could you elaborate more about it? We can make a short call as well.

File: instant-xml/src/lib.rs
Comment (djc) [reply instant-xml/src/lib.rs]: We should probably support tuple structs, but not with scalars (directly) inside them.

File: instant-xml/src/lib.rs
Comment (djc) [reply instant-xml/src/lib.rs]: Let's make it immutable for now, can always revisit the decision later.

File: instant-xml/src/lib.rs
Comment (djc) [reply instant-xml/src/lib.rs:14]: If we have a generic `W: fmt::Write`, callers will monomorphize the implementations with a particular `fmt::Write` implementations. If we instead hold a `&mut dyn fmt::Write` in the `Serializer`, each call to write some data into the `fmt::Write` will have to go through the method table used for the `&dyn` fat pointer. Given that we're doing a lot of calls into that, the performance implications for that start to add up.

File: instant-xml-macros/src/lib.rs
Comment (djc) [reply instant-xml-macros/src/lib.rs:234]: Okay, fair enough. I guess my main concern was allocating a `Vec` for this, I guess we could just create an array here instead (since we know the length of the array at codegen time).

Comment (choinskib) [reply instant-xml-macros/src/lib.rs:234]: We don't, as we don't know which prefixes already exist in parent. We are checking it in the next line.

Comment (djc) [reply instant-xml-macros/src/lib.rs:234]: Okay, never mind, we can optimize this later.

**Review by djc — APPROVED** [review]

**Pull request merged**
